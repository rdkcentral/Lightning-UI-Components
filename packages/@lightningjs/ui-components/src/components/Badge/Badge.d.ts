/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import lng from '@lightningjs/core';
import Base from '../Base';
import { Color, StylePartial } from '../../types/lui';

type BadgeStyle = {
  backgroundColor: Color;
  contentSpacing: number;
  iconColor: Color;
  offsetY: number;
  paddingX: number;
  paddingY: number;
  radius: lng.Tools.CornerRadius;
  strokeColor: Color;
  strokeWidth: number;
  textStyle: lng.textures.TextTexture.Settings;
};

declare namespace Badge {
  export interface TemplateSpec extends Base.TemplateSpec {
    /**
     * path to image or inline SVG XML
     */
    icon?: string;

    /**
     * width of icon
     */
    iconWidth?: number;

    /**
     * height of icon
     */
    iconHeight?: number;

    /**
     * side of the text where icon will appear on (`left` or `right`)
     */
    iconAlign?: 'left' | 'right';

    /**
     * Badge text
     */
    title?: string;
  }

  export interface TypeConfig extends lng.Component.TypeConfig {
    SignalMapType: SignalMap;
  }
  /**
   * emitted when the Badge component finishes loading its content and adjusts the background to fit the text input size.
   */
  export type SignalMap = {
    loadedBadge(): void;
  };
}

declare class Badge<
  TemplateSpec extends Badge.TemplateSpec = Badge.TemplateSpec,
  TypeConfig extends Badge.TypeConfig = Badge.TypeConfig
> extends Base<TemplateSpec, TypeConfig> {
  /**
   * path to image or inline SVG XML
   */
  icon?: string;

  /**
   * width of icon
   */
  iconWidth?: number;

  /**
   * height of icon
   */
  iconHeight?: number;

  /**
   * side of the text where icon will appear on (`left` or `right`)
   */
  iconAlign?: 'left' | 'right';

  /**
   * Badge text
   */
  title?: string;

  get style(): BadgeStyle;
  set style(v: StylePartial<BadgeStyle>);

  // tags
  get _Background(): lng.Component;
  get _Text(): lng.Component;
  get _Icon(): lng.Component;
}

export { Badge as default, BadgeStyle };
