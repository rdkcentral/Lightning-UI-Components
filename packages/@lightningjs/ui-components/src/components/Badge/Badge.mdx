import { Meta, Canvas } from '@storybook/blocks';
import Badge from '.';
import * as BadgeStories from './Badge.stories.js';

<Meta of={BadgeStories} />

# Badge

This component is used to display small keywords of metadata, like video or audio quality (ie HD or CC).

## Source

https://github.com/rdkcentral/Lightning-UI-Components/blob/develop/packages/@lightningjs/ui-components/src/components/Badge/Badge.js

## Usage

The basic component just requires a title:

```js
import { Badge } from '@lightningjs/ui-components';
import circle from '../../assets/images/circle.svg';

class Example extends lng.Component {
  static _template() {
    return {
      Badge: {
        type: Badge,
        title: 'HD'
      }
    };
  }
}
```

<<<<<<< HEAD

<Canvas>
  <Story id="components-badge--text" />
</Canvas>

=======

> > > > > > > 5e82b56b (feature(Storybook): updates mdx files with proper imports and renames doc only mdx files)
> > > > > > > Or an icon:

```js
import { Badge } from '@lightningjs/ui-components';

class Example extends lng.Component {
  static _template() {
    return {
      Badge: {
        type: Badge,
        icon: circle,
        iconWidth: 50,
        iconHeight: 50
      }
    };
  }
}
```

<<<<<<< HEAD

<Canvas>
  <Story id="components-badge--icon-png" />
</Canvas>

=======

> > > > > > > 5e82b56b (feature(Storybook): updates mdx files with proper imports and renames doc only mdx files)
> > > > > > > A title and an icon can also be used together:

```js
import { Badge } from '@lightningjs/ui-components';
import lightning from '../../assets/images/ic_lightning_white_32.png';

class Example extends lng.Component {
  static _template() {
    return {
      Badge: {
        type: Badge,
        title: 'HD',
        icon: lightning
      }
    };
  }
}
```

## API

### Properties

| name       | type   | required | default   | description                                                    |
| ---------- | ------ | -------- | --------- | -------------------------------------------------------------- |
| icon       | string | false    | undefined | path to image or inline SVG XML                                |
| iconWidth  | number | false    | undefined | width of icon (required for SVG icons)                         |
| iconHeight | number | false    | undefined | height of icon (required for SVG icons)                        |
| iconAlign  | string | false    | undefined | side of the text where icon will appear on (`left` or `right`) |
| title      | string | false    | undefined | badge text                                                     |

### Style Properties

| name            | type             | description                                                          |
| --------------- | ---------------- | -------------------------------------------------------------------- |
| backgroundColor | string           | color of the background                                              |
| contentSpacing  | number           | padding to add between the icon and the text                         |
| iconColor       | string           | color to be applied to the icon (only works with pure white PNGs)    |
| offsetY         | number           | offset used to vertically center the text within the container       |
| paddingX        | number           | padding to add on the left and right between the text and background |
| paddingY        | number           | padding to add on the top and bottom between the text and background |
| radius          | number \| array  | radius of badge corner                                               |
| strokeColor     | string           | color of the Badge border                                            |
| strokeWidth     | number           | width of the Badge border                                            |
| textStyle       | string \| object | text style to apply to the title                                     |

### Events

#### $loadedBadge():void

This event is fired when the contents of the badge have finished loading and the background has been appropriately sized based on the text input.
