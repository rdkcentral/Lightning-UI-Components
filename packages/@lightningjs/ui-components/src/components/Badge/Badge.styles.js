/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

export const base = theme => {
  return {
    contentSpacing: theme.spacer.sm,
    // offsetY: 1, //TODO needto confirm before if I can remove offSet from Badge Styles
    paddingX: theme.spacer.md + theme.spacer.xxs,
    paddingY: theme.spacer.xs,
    radius: theme.radius.sm,
    strokeWidth: theme.stroke.sm,
    textStyle: {
      ...theme.typography.tag1,
      textAlign: 'center'
    }
  };
};

export const tone = theme => ({
  neutral: {
    backgroundColor: theme.color.fillInverseSecondary,
    iconColor: theme.color.textNeutral,
    textStyle: { textColor: theme.color.textNeutral },
    strokeColor: theme.color.strokeNeutral
  },
  inverse: {
    backgroundColor: theme.color.fillNeutralSecondary,
    iconColor: theme.color.textInverse,
    textStyle: { textColor: theme.color.textInverse },
    strokeColor: theme.color.strokeInverse
  },
  brand: {
    backgroundColor: theme.color.fillBrand,
    iconColor: theme.color.textNeutral,
    textStyle: { textColor: theme.color.textNeutral },
    strokeColor: theme.color.strokeInverse
  }
});
