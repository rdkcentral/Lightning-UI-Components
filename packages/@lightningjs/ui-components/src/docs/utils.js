/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

// eslint-disable-next-line no-unused-vars
import React from 'react';
import utils from '../utils';

// creates Mode control on stories
export const createModeControl = ({ options, summaryValue } = {}) => {
  return {
    mode: {
      control: 'radio',
      options:
        options && Array.isArray(options)
          ? ['unfocused', ...options]
          : ['unfocused', 'focused', 'disabled'],
      description: 'Sets the visual mode for the component',
      table: {
        defaultValue: { summary: summaryValue }
      },
      type: { name: 'string', required: true }
    }
  };
};

/**
 * @param {object} config Parameters for generating nested args: { argsObj, targetProp, include, overrides }
 */
export function nestedArgs(config) {
  const { argsObj, targetProp, include, overrides } = config;
  return Object.keys(argsObj).reduce((acc, curr) => {
    if (!include.includes(curr)) return acc;
    return {
      ...acc,
      [targetProp + '_' + curr]:
        (overrides.args && overrides.args[curr]) || argsObj[curr]
    };
  }, {});
}

/**
 * @param {object} config Parameters for generating nested arg types: { argsObj, targetProp, include, overrides }
 */
export function nestedArgTypes(config) {
  const { argTypesObj, targetProp, include, overrides } = config;
  return Object.keys(argTypesObj).reduce((acc, curr) => {
    if (!include.includes(curr)) return acc;
    return {
      ...acc,
      [targetProp + '_' + curr]: {
        // Namespaced to avoid conflicts
        name: curr,
        ...argTypesObj[curr],
        ...overrides[curr],
        table: {
          ...(argTypesObj[curr].table || {}),
          ...(overrides[curr] && overrides[curr].table),
          category: targetProp
        }
      }
    };
  }, {});
}

export const prevValues = {};

/**
 * @param {object} config Parameters for generating nested arg actions: { componentName, argsTypeObj, targetProp, include, overrides }
 */
export function nestedArgActions(config) {
  const { componentName, argTypesObj, targetProp, include, overrides } = config;
  return Object.keys(argTypesObj).reduce((acc, curr) => {
    if (!include.includes(curr)) return acc;
    return {
      ...acc,
      [targetProp + '_' + curr]: (value, component) => {
        component.tag(componentName).patch({
          [targetProp]: {
            ...prevValues[targetProp],
            [curr.replace(targetProp + '_', '')]:
              value === 'none'
                ? undefined
                : (overrides && overrides[curr]) || value // There are issues with merging objects here
          }
        });
        // Allow patching to work with nested objects
        if (!prevValues[targetProp]) {
          prevValues[targetProp] = {};
        }
        prevValues[targetProp][curr.replace(targetProp + '_', '')] =
          value === 'none'
            ? undefined
            : (overrides && overrides[curr]) || value;
      }
    };
  }, {});
}

/**
 * @param {object} config Parameters for generating story: { componentName, baseStory, subStory, targetProperty, include, options }
 */
export function generateSubStory(config) {
  config.baseStory.args = {
    ...config.baseStory.args,
    ...nestedArgs({
      argsObj: config.subStory.args || {},
      targetProp: config.targetProperty,
      include: config.include,
      overrides: (config.overrides && config.overrides.args) || {}
    })
  };

  config.baseStory.argTypes = {
    ...config.baseStory.argTypes,
    ...nestedArgTypes({
      argTypesObj: config.subStory.argTypes || {},
      targetProp: config.targetProperty,
      include: config.include,
      overrides: (config.overrides && config.overrides.argTypes) || {}
    })
  };

  if (
    !(
      config.baseStory &&
      config.baseStory.parameters &&
      config.baseStory.parameters.argActions
    )
  ) {
    config.baseStory.parameters = {
      argActions: {}
    };
  }

  config.baseStory.parameters.argActions = {
    ...config.baseStory.parameters.argActions,
    ...nestedArgActions({
      componentName: config.componentName,
      argTypesObj: config.subStory.argTypes || {},
      targetProp: config.targetProperty,
      include: config.include,
      overrides: (config.overrides && config.overrides.argActions) || {}
    })
  };
}

// used to create links to Lightning docs given an id
export const DocsLink = ({ children, id }) => {
  const docsmap = {
    lng: {
      _base: 'https://rdkcentral.github.io/Lightning/docs',
      Component: '/components/overview',
      Text: '/textures/text',
      Transition: '/transitions/attributes'
    }
  };
  const [source, path] = id.split('.');
  const href = docsmap[source]._base + docsmap[source][path];
  return (
    <a href={href} target="_blank">
      {children || id}
    </a>
  );
};

export const Item = ({ children }) => children;

/**
 * Creates formatted object for markdown prop tables
 * see Column and Row stories
 * @param {object} props
 * @param {object} props.type - TS-style generic/utility type
 */
export const ObjectFormat = ({ object }) => {
  return [
    '{',
    <br />,
    '  ',
    ...utils.flatten(
      Object.entries(object).map(([k, v]) => [
        <span>&nbsp;&nbsp;</span>,
        `${k}: `,
        v,
        ',',
        <br />
      ])
    ),
    '}'
  ].map((x, i) => <Item key={i}>{x}</Item>);
};

/**
 * Formats TS-style generic types
 * See Columns and Row stories
 * @param {object} props
 * @param {object} props.children - component children
 * @param {string} props.type - Name of generic type
 */
export const GenericType = ({ children, type }) => {
  return [type, '<', children, '>'].map((x, i) => <Item key={i}>{x}</Item>);
};
