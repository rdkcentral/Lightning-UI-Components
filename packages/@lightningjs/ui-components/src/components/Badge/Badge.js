/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import Base from '../Base';
import Icon from '../Icon';
import { getHexColor } from '../../utils';
import * as styles from './Badge.styles.js';
import lng from '@lightningjs/core';

export default class Badge extends Base {
  static _template() {
    const center = {
      mount: 0.5,
      x: w => w / 2,
      y: h => h / 2
    };
    return {
      rect: true,
      shader: {
        type: lng.shaders.RoundedRectangle
      },
      Text: {
        mountY: 0.5,
        text: {}
      },
      Icon: {
        type: Icon,
        mountY: 0.5,
        signals: {
          itemChanged: '_updateLayout'
        }
      },
      Stroke: center
    };
  }

  static get __componentName() {
    return 'Badge';
  }

  static get __themeStyle() {
    return styles;
  }

  static get properties() {
    return ['title', 'icon', 'iconAlign', 'iconWidth', 'iconHeight'];
  }

  static get tags() {
    return ['Text', 'Icon', 'Stroke'];
  }

  _init() {
    this._Text.on('txLoaded', this._updateLayout.bind(this));
    super._init();
  }

  _update() {
    this._updateText();
    this._updateStroke();
    this._updateIcon();
    this._updateLayout();
  }

  _updateLayout() {
    this._updateWidth();
    this._updateBackground();
    this._updatePositions();
  }

  _updateBackground() {
    this.patch({
      h:
        Math.max(this._Text.renderHeight, this._Icon.h) +
        this.style.paddingY * 2,
      color: this.style.backgroundColor,
      shader: { radius: this.style.radius }
    });
  }

  _updateText() {
    if (this._Text) {
      this._Text.patch({
        text: {
          ...this.style.textStyle,
          text: this.title || ''
        }
      });
    }
  }

  _updateStroke() {
    this._Stroke.patch({
      texture: lng.Tools.getRoundRect(
        this.w,
        this.h,
        this.style.radius,
        this.style.strokeWidth,
        this.style.strokeColor,
        false
      )
    });
  }

  _updateIcon() {
    this._Icon.patch({
      icon: this.icon,
      w: this.iconWidth,
      h: this.iconHeight,
      style: {
        color: getHexColor(this.style.iconColor)
      }
    });
  }

  _updateWidth() {
    let contentSpacing = 0;
    if (this.icon && this.title) {
      contentSpacing = this.style.contentSpacing;
    }

    this.w = this.title
      ? this._Text.renderWidth +
        this.style.paddingX * 2 +
        (this._Icon.finalW || 0) +
        contentSpacing
      : this.style.paddingX * 2 + (this._Icon.finalW || 0);
  }

  _updatePositions() {
    this._Icon.y = this.h / 2;

    if (this.iconAlign === 'left' && this.title) {
      this._Icon.x = this.style.paddingX;
      this._Text.x =
        this._Icon.x + this._Icon.finalW + this.style.contentSpacing;
    } else if (this.iconAlign === 'right' && this.title) {
      this._Text.x = this.style.paddingX;
      this._Icon.x =
        this._Text.x + this._Text.renderWidth + this.style.contentSpacing;
    } else {
      this._Text.x = this.style.paddingX;
      this._Icon.x = this.style.paddingX;
    }

    this._Text.y = this._h / 2 + this.style.offsetY;
    this.signal('loadedBadge', this);
  }

  _getIconHeight() {
    if (this.icon) {
      return !this._Icon.finalH && this._Text
        ? this._Text.text.lineHeight
        : this._Icon.finalH;
    }
    return 0;
  }

  set announce(announce) {
    super.announce = announce;
  }

  get announce() {
    return this._announce || (this._Text && this._Text.text.text);
  }
}
