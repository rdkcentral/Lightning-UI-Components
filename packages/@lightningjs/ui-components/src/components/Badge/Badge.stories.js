/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import lng from '@lightningjs/core';
import Badge from '.';
import mdx from './Badge.mdx';
import lightning from '../../assets/images/ic_lightning_white_32.png';
import circle from '../../assets/images/circle.svg';
import { CATEGORIES } from '../../docs/constants';

export default {
  title: `${CATEGORIES[32]}/Badge`,
  parameters: {
    docs: {
      page: mdx
    }
  }
};

export const Text = args =>
  class Basic extends lng.Component {
    static _template() {
      return {
        Badge: {
          type: Badge,
          title: args.title
        }
      };
    }
  };
Text.args = {
  title: 'HD'
};
Text.argTypes = {
  title: {
    description: 'Badge text',
    control: 'text',
    table: {
      defaultValue: { summary: 'undefined' }
    }
  }
};

export const IconSVG = () =>
  class Basic extends lng.Component {
    static _template() {
      return {
        Badge: {
          type: Badge,
          icon: circle,
          iconWidth: 24,
          title: 'HD'
        }
      };
    }
  };
IconSVG.args = {
  iconAlign: 'left'
};

IconSVG.argTypes = {
  iconAlign: {
    control: 'select',
    options: ['left', 'right'],
    description: 'Side of the text the icon will appear on',
    table: {
      defaultValue: { summary: 'undefined' }
    }
  }
};
export const IconPNG = () =>
  class Basic extends lng.Component {
    static _template() {
      return {
        Badge: {
          type: Badge,
          icon: lightning
        }
      };
    }
  };

export const TextWithIcon = () =>
  class Basic extends lng.Component {
    static _template() {
      return {
        Badge: {
          type: Badge,
          icon: lightning
        }
      };
    }
  };

TextWithIcon.args = {
  title: 'HD',
  iconAlign: 'left'
};

TextWithIcon.argTypes = {
  iconAlign: {
    control: 'select',
    options: ['left', 'right'],
    description: 'Side of the text the icon will appear on',
    table: {
      defaultValue: { summary: 'undefined' }
    }
  },
  title: {
    control: 'text',
    description: 'Badge text',
    table: {
      defaultValue: { summary: 'undefined' }
    }
  }
};
