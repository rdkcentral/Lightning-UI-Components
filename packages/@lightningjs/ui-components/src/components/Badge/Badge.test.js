/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  makeCreateComponent,
  pathToDataURI
} from '@lightningjs/ui-components-test-utils';
import Badge from '.';
import { getHexColor } from '../../utils';

const createBadge = makeCreateComponent(Badge);

const circleIcon = pathToDataURI('src/assets/images/circle.svg');
const xfinityLogo = pathToDataURI(
  'src/assets/images/Xfinity-Provider-Logo-Square.png'
);

describe('Badge', () => {
  let badge, testRenderer;

  beforeEach(() => {
    [badge, testRenderer] = createBadge();
  });

  afterEach(() => {
    badge = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('sets the announce string to the badge text', () => {
    const title = 'HD';
    badge.title = title;
    testRenderer.forceAllUpdates();
    expect(badge.announce).toBe(title);
  });

  it('overrides the announce string', () => {
    const overrideString = 'Custom announce string';
    badge.announce = overrideString;
    testRenderer.forceAllUpdates();
    expect(badge.announce).toBe(overrideString);
  });

  it('renders icon and text side by side', () => {
    const [badge] = createBadge({
      title: 'HD',
      icon: circleIcon
    });
    expect(badge).toBeDefined();
  });

  it('renders icon only', () => {
    const [badge] = createBadge({
      icon: circleIcon
    });
    expect(badge).toBeDefined();
  });

  it('should update the title', async () => {
    const title = 'HD';
    [badge, testRenderer] = createBadge(
      { title },
      { spyOnMethods: ['_update', '_updateBackground'] }
    );
    setTimeout(() => {
      badge._badgeTextLoaded();
    }, 500);
    await badge.__updateBackgroundSpyPromise;
    testRenderer.forceAllUpdates();
    expect(badge.title).toBe(title);
    expect(badge.w).toBe(
      badge._Text.renderWidth +
        badge.style.offsetX * 2 +
        (badge._Icon.finalW || 0)
    );

    const title2 = 'longer title';
    badge.title = title2;
    testRenderer.forceAllUpdates();
    setTimeout(() => {
      badge._badgeTextLoaded();
    }, 500);
    await badge.__updateBackgroundSpyPromise;
    expect(badge.title).toBe(title2);
    expect(badge.w).toBe(
      badge._Text.renderWidth +
        badge.style.offsetX * 2 +
        (badge._Icon.finalW || 0)
    );
    expect(badge.h).toBe(
      Math.max(badge._Text.renderHeight, badge._Icon.h) +
        badge.style.paddingY * 2
    );
  });

  it('should update the background', () => {
    const color = getHexColor('ffffff', 0.25);
    [badge] = createBadge({ background: { color }, icon: undefined });
    expect(badge.background.color).toBe(color);
  });

  it('should update the text style', async () => {
    const color = getHexColor('ffffff', 0.25);
    const title = 'HD';
    const icon = xfinityLogo;
    const textProperties = { color, textAlign: 'left' };

    [badge, testRenderer] = createBadge(
      {
        title,
        icon,
        h: 30,
        iconAlign: 'right',
        iconWidth: 50,
        iconHeight: 50
      },
      { spyOnMethods: ['_updateBackground'] }
    );
    badge.style = { paddingX: 0 };
    testRenderer.forceAllUpdates();
    setTimeout(() => {
      badge._badgeTextLoaded();
      badge._iconLoaded();
    }, 500);
    await badge.__updateBackgroundSpyPromise;
    badge._Text.text = textProperties;
    expect(badge._Text.x).toBe(0);
    expect(badge._Text.y).toBe(badge.h / 2 + 2);
    // This should work once bug in TextBox is fixed to patch textStyle
    expect(badge._Text.text.textAlign).toBe(textProperties.textAlign);
    expect(badge._Text.text.color).toBe(color);
  });

  it('should update the icon', () => {
    const icon = xfinityLogo;
    [badge, testRenderer] = createBadge({ icon });
    expect(badge.icon).toBe(icon);
    const icon2 = circleIcon;
    badge.icon = icon2;
    expect(badge.icon).toBe(icon2);
    expect(badge._Icon.color).toBe(badge._Text.color);
  });

  it('should position the text and icon based on the iconAlign property (right)', async () => {
    const title = 'HD';
    const icon = xfinityLogo;
    [badge, testRenderer] = createBadge(
      {
        title,
        icon,
        iconAlign: 'right',
        iconWidth: 50,
        iconHeight: 50
      },
      {
        spyOnMethods: ['_updateBackground', '_updateIcon']
      }
    );
    setTimeout(() => {
      badge._badgeTextLoaded();
      badge._iconLoaded();
    }, 500);
    await badge.__updateBackgroundSpyPromise;

    expect(badge._Text.x).toEqual(badge.style.paddingX);
    expect(badge._Icon.x).toEqual(
      badge._Text.x + badge._Text.renderWidth + badge.style.contentSpacing
    );
  });

  it('should position the text and icon based on the iconAlign property (left)', async () => {
    const title = 'HD';
    const icon = xfinityLogo;
    [badge, testRenderer] = createBadge(
      {
        title,
        icon,
        iconAlign: 'left',
        iconWidth: 50,
        iconHeight: 50
      },
      {
        spyOnMethods: ['_updateBackground', '_updateIcon']
      }
    );
    setTimeout(() => {
      badge._badgeTextLoaded();
      badge._iconLoaded();
    }, 500);
    await badge.__updateBackgroundSpyPromise;
    expect(badge._Text.x).toEqual(
      badge._Icon.x + badge._Icon.finalW + badge.style.contentSpacing
    );
    expect(badge._Icon.x).toEqual(badge.style.paddingX);
  });
});
