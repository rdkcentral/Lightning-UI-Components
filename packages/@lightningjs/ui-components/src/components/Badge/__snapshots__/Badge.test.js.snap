// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge renders 1`] = `
{
  "Component": {
    "Icon": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294506490,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "hasFinalFocus": false,
      "hasFocus": false,
      "isComponent": true,
      "mount": 0,
      "mountX": 0,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Icon",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": "",
      "tag": [Function],
      "tags": [
        "Icon",
      ],
      "type": "Icon",
      "visible": true,
      "w": 0,
      "x": 12,
      "y": 4,
      "zIndex": 0,
    },
    "Text": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Text",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Text",
      ],
      "texture": {
        "fontFace": "Arial",
        "fontSize": 20,
        "fontStyle": "500",
        "lineHeight": 24,
        "precision": 0.6666666666666666,
        "textAlign": "center",
        "textBaseline": "bottom",
        "textColor": 4294506490,
        "type": "TextTexture",
        "verticalAlign": "middle",
      },
      "visible": true,
      "w": 0,
      "x": 12,
      "y": 8,
      "zIndex": 0,
    },
  },
}
`;
